bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieProtoEncoding.java

```diff
@@ -260,6 +298,26 @@ public Object encode(Object msg, ByteBufAllocator allocator)
                     } else {
                         return ByteBufList.get(buf, rr.getData());
                     }
+                }  else if (msg instanceof BookieProtocol.BatchedReadResponse) {
+                    BookieProtocol.BatchedReadResponse brr = (BookieProtocol.BatchedReadResponse) r;
+                    int payloadSize = brr.getData().readableBytes();
+                    int delimiterSize = brr.getData().size() * 4; // The size of each entry.
+                    int responseSize = RESPONSE_HEADERS_SIZE + 8 /* request_id */ + payloadSize + delimiterSize;
+                    int bufferSize = 4 /* frame size */ + responseSize;
+                    ByteBuf buf = allocator.buffer(bufferSize);
+                    buf.writeInt(responseSize);
+                    buf.writeInt(PacketHeader.toInt(r.getProtocolVersion(), r.getOpCode(), (short) 0));
+                    buf.writeInt(r.getErrorCode());
+                    buf.writeLong(r.getLedgerId());
+                    buf.writeLong(r.getEntryId());
+                    buf.writeLong(((BookieProtocol.BatchedReadResponse) r).getRequestId());
+                    for (int i = 0; i < brr.getData().size(); i++) {
+                        ByteBuf entryData = brr.getData().getBuffer(i);
+                        buf.writeInt(entryData.readableBytes());
+                        buf.writeBytes(entryData);
```

Comment:
If the `payloadSize` is less than SMALL_ENTRY_SIZE_THRESHOLD, we can use `buf.writeBytes(entryData)` to copy byebuf data. Otherwise, we can use ByteBufList to avoid data copy.

Reply:
addressed.